lib/Core/Executor.cpp

```diff
@@ -2548,6 +2548,159 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
     // instructions.
     terminateStateOnExecError(state, "Unexpected ShuffleVector instruction");
     break;
+
+  case Instruction::AtomicRMW: {
+    // An atomic instruction gets a pointer and a value. It reads the value at the pointer, perfoms its operation, stores the result and returns the value that was originally at the pointer.
+    const AtomicRMWInst *ai = cast<AtomicRMWInst>(i);
+    switch (ai->getOperation()) {
+    case AtomicRMWInst::Xchg: {
+      ref<Expr> pointer = eval(ki, 0, state).value;
+      ref<Expr> value = eval(ki, 1, state).value;
+
+      executeMemoryOperation(state, false, pointer, 0, ki);
+
+      executeMemoryOperation(state, true, pointer, value, 0);
+      break;
+    }
+    case AtomicRMWInst::Add: {
+      ref<Expr> pointer = eval(ki, 0, state).value;
+      ref<Expr> value = eval(ki, 1, state).value;
+
+      executeMemoryOperation(state, false, pointer, 0, ki);
+      ref<Expr> oldValue = getDestCell(state, ki).value;
+
+      ref<Expr> result = AddExpr::create(oldValue, value);
```

Comment:
For example, here I think it's better that `result` is a `ref<Expr>` than a `ref<AddExpr>`, because it communicates the intention that `result` can be any old `ref<Expr>`

Reply:
The point of `auto` is that you don't write this. The `result` becomes an argument of `executeMemoryOperation`, you don't need anything inbetween the variable name conveys the message.